kratos/containers/pointer_vector_set.h

```diff
@@ -1066,51 +1066,52 @@ class PointerVectorSet final
             return;
         }
 
-        if (empty()) {
-            mData.reserve(std::distance(first, last));
+        if (empty() || (KeyOf(end() - 1) < KeyOf(GetReference(first)))) {
+            mData.reserve(mData.size() + std::distance(first, last));
             for (auto it = first; it != last; ++it) {
                 mData.push_back(TPointerType(&GetReference(it)));
             }
         } else {
-            // first find the largest range
-            const auto lower_bound_first = std::lower_bound(mData.begin(), mData.end(), KeyOf(GetReference(first)), CompareKey());
-            const auto upper_bound_last = std::upper_bound(lower_bound_first, mData.end(), KeyOf(GetReference(last-1)), CompareKey());
-
-            // then find the compact sub range
-            const auto upper_bound_first = std::upper_bound(lower_bound_first, upper_bound_last, KeyOf(GetReference(first)), CompareKey());
-            const auto lower_bound_last = std::lower_bound(lower_bound_first, upper_bound_last, KeyOf(GetReference(last-1)), CompareKey());
-
-            if (lower_bound_first == lower_bound_last &&
-                lower_bound_first == upper_bound_first &&
-                lower_bound_first == upper_bound_last)
-            {
-                // all 4 bounds are equal, hence this can be inserted without checking further
-                mData.reserve(mData.size() + std::distance(first, last));
-                if (lower_bound_first == mData.end()) {
-                    for (auto it = first; it != last; ++it) {
-                        mData.push_back(TPointerType(&GetReference(it)));
-                    }
+            // rather than using insert which always needs to move, we push_back to a new container,
+            // and then swap the underlying data.
+            TContainerType aux_container;
+            aux_container.reserve(mData.size() + std::distance(first, last));
+
+            auto existing_data_it = mData.begin();
+            auto new_data_it = first;
+
+            while (existing_data_it != mData.end() && new_data_it != last) {
+                const auto existing_key = KeyOf(GetReference(existing_data_it));
+                const auto new_key = KeyOf(GetReference(new_data_it));
+                if (existing_key == new_key) {
+                    // adds the existing item and ignore the new item to keep consistent with the std::set::insert
+                    aux_container.push_back(TPointerType(&GetReference(existing_data_it)));
```

Comment:
If the key exist we should simply replace the pointers, not creating a new one

Reply:
This is done like this because, now we support different types of iterators (`std::vector`, ...). So we need to bring everything to a common base. I understand what you mean. In this case, I would implement `GetPointer` method, such that it will return the same pointer if the iterator is having a pointer type, otherwise it will create a pointer and return/move it.